libs/dyn/pattern_string.go

```diff
@@ -0,0 +1,102 @@
+package dyn
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+)
+
+// MustPatternFromString is like NewPatternFromString but panics on error.
+func MustPatternFromString(input string) Pattern {
+	p, err := NewPatternFromString(input)
+	if err != nil {
+		panic(err)
+	}
+	return p
+}
+
+// NewPatternFromString parses a pattern from a string.
+//
+// The string must be a sequence of keys and indices separated by dots.
+// Indices must be enclosed in square brackets.
+// The string may include a leading dot.
+// The wildcard character '*' can be used to match any key or index.
+//
+// Examples:
+//   - foo.bar
+//   - foo[1].bar
+//   - foo.*.bar
+//   - foo[*].bar
+//   - .
+func NewPatternFromString(input string) (Pattern, error) {
+	var pattern Pattern
+
+	p := input
+
+	// Trim leading dot.
+	if p != "" && p[0] == '.' {
+		p = p[1:]
+	}
+
+	for p != "" {
+		// Every component may have a leading dot.
+		if p != "" && p[0] == '.' {
```

Comment:
I think you can omit `p != ""` from here

Reply:
You're right -- this was copied from the path equivalent of the same function.

I must have copied the condition from before the for loop.

Reply:
Thanks dropped this in both path_string.go and patterns_string.go. And yes, pattern_string.go follows path_string.go very closely.